src/Microsoft.Extensions.DependencyInjection.Abstractions/ServiceProviderFactory.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Provides an abstract base for creating a container specific builder and an <see cref="IServiceProvider"/>.
+    /// </summary>
+    public abstract class ServiceProviderFactory<TContainerBuilder> : IServiceProviderFactory
+    {
+        /// <summary>
+        /// Creates a container builder from an <see cref="IServiceCollection"/>.
+        /// </summary>
+        /// <param name="services">The collection of services</param>
+        /// <returns>A container builder that can be used to create an <see cref="IServiceProvider"/>.</returns>
+        public abstract TContainerBuilder CreateBuilder(IServiceCollection services);
```

Comment:
`TBuilder`? How are you supposed to get to this type from `IServiceProvider`? Right now you have to call this method and do `builder.GetType()` in order to get it, but that might not be the exact type you'd want. Say you have a `ServiceProviderFactory<IContainerBuilder>`, but the actual type is `ContainerBuilder`. How will you find the `ConfigureContainer` method? Check for the parameter type using `IsAssignableFrom`?